scripts/deep/model.py

```diff
@@ -1,60 +1,264 @@
 
-        Parameters:
-            X (pd.DataFrame): dataframe of user and recipe ids for which to predict ratings
+    model = RecipeRecommenderGNN(metadata=data.metadata()).to(device)
+    optimizer = optim.Adam(model.parameters(), lr=1e-4)
 
+    EPOCHS = 10
+    print("\nStarting training...\n")
+
+    for epoch in range(1, EPOCHS + 1):
+        model.train()
+        optimizer.zero_grad()
+        start = time.time()
+
+        # Construct input features
+        x_dict = {}
+        for node_type in data.metadata()[0]:
+            if node_type == 'recipe':
+                x_dict[node_type] = data[node_type].x
+            else:
+                x_dict[node_type] = torch.arange(data.num_nodes_dict[node_type], device=device)
+
+        # Forward pass
+        out_dict = model(x_dict, data.edge_index_dict, data.num_nodes_dict)
+
+        # Get positive and negative samples
+        pos_u, pos_r, neg_u, neg_r = get_pos_neg_edges(data)
+
+        pos_scores = model.predict(out_dict['user'][pos_u], out_dict['recipe'][pos_r])
+        neg_scores = model.predict(out_dict['user'][neg_u], out_dict['recipe'][neg_r])
+
+        # Create binary labels and compute loss
+        labels = torch.cat([torch.ones_like(pos_scores), torch.zeros_like(neg_scores)])
+        scores = torch.cat([pos_scores, neg_scores])
+
+        loss = F.binary_cross_entropy_with_logits(scores, labels)
+        loss.backward()
+        optimizer.step()
+
+        print(f"Epoch {epoch}/{EPOCHS} — Loss: {loss.item():.4f} — Time: {time.time() - start:.2f}s")
+
+    print("\nTraining complete!")
+
+    # === Export model and embeddings ===
+    print("Exporting model...")
+    model.eval()
+    x_dict = {
+        node_type: (
+            data[node_type].x if node_type == 'recipe'
+            else torch.arange(data.num_nodes_dict[node_type], device=device)
+        )
+        for node_type in data.metadata()[0]
+    }
+
+    final_embs = model(x_dict, data.edge_index_dict, data.num_nodes_dict)
+    export_script_model(model, final_embs)
+
+    # Save raw PyTorch checkpoint
+    torch.save(model.state_dict(), "recipe_model.pt")
+
+
+class RecipeGNNModel:
+    """
+    Wrapper for evaluating a trained GNN model using internal index-based test data.
+    """
+    def __init__(self, model, data, device):
+        self.model = model.eval()
+        self.data = data
+        self.device = device
+        self.out_dict = self._generate_embeddings()
+
+    def _generate_embeddings(self):
+        """
+        Generate final embeddings for all nodes in the graph.
+        """
+        x_dict = {
+            node_type: (
+                self.data[node_type].x if node_type == 'recipe'
+                else torch.arange(self.data.num_nodes_dict[node_type], device=self.device)
+            )
+            for node_type in self.data.metadata()[0]
+        }
+
+        return self.model(x_dict, self.data.edge_index_dict, self.data.num_nodes_dict)
+
+    def evaluate_from_index(self, test_df):
```

Comment:
Where are you calling this function? Can you include the script to run this evaluation and also report the evaluation result in the PR description/conversation?